File: core-dna/mcp-server.py
#!/usr/bin/env python3
"""
Zircon MCP Server — 标准 MCP 协议
任何支持 MCP 的 LLM 都能调用 Zircon 的能力
"""

import json
import sys
from pathlib import Path

CORE_DIR = Path("/home/.openclaw/workspace/core-dna")
MEMORY_DIR = Path("/root/.openclaw/workspace/memory")

# === MCP 工具定义 ===
TOOLS = [
    {
        "name": "zircon_status",
        "description": "获取 Zircon 系统状态",
        "inputSchema": {"type": "object", "properties": {}}
    },
    {
        "name": "zircon_genes",
        "description": "获取基因库状态",
        "inputSchema": {"type": "object", "properties": {}}
    },
    {
        "name": "zircon_mutate",
        "description": "执行基因变异",
        "inputSchema": {
            "type": "object",
            "properties": {
                "domain": {"type": "string", "description": "变异维度"},
                "change": {"type": "number", "description": "变异幅度"}
            },
            "required": ["domain", "change"]
        }
    },
    {
        "name": "zircon_echo",
        "description": "运行回音壁知识增强",
        "inputSchema": {"type": "object", "properties": {}}
    },
    {
        "name": "zircon_evolve",
        "description": "运行一轮进化",
        "inputSchema": {"type": "object", "properties": {}}
    },
]

# === MCP 响应 ===
def mcp_response(req_id, result):
    return json.dumps({"jsonrpc": "2.0", "id": req_id, "result": result})

def mcp_error(req_id, code, message):
    return json.dumps({"jsonrpc": "2.0", "id": req_id, "error": {"code": code, "message": message}})

# === 工具实现 ===
def handle_tool(name, args):
    if name == "zircon_status":
        state_file = MEMORY_DIR / "daemon-state.json"
        if state_file.exists():
            with open(state_file) as f:
                return {"content": [{"type": "text", "text": json.dumps(json.load(f), indent=2)}]}
        return {"content": [{"type": "text", "text": "守护进程未运行"}]}
    
    elif name == "zircon_genes":
        gene_file = MEMORY_DIR / "gene-registry.json"
        if gene_file.exists():
            with open(gene_file) as f:
                r = json.load(f)
            genes = r.get("genes", [])
            cats = {}
            for g in genes:
                c = g.get("category", "?")
                cats[c] = cats.get(c, 0) + 1
            summary = f"总基因: {len(genes)}\n分布: {json.dumps(cats, ensure_ascii=False)}"
            return {"content": [{"type": "text", "text": summary}]}
        return {"content": [{"type": "text", "text": "基因库不存在"}]}
    
    elif name == "zircon_mutate":
        return {"content": [{"type": "text", "text": f"变异: {args.get('domain')} ({args.get('change')})"}]}
    
    elif name == "zircon_echo":
        return {"content": [{"type": "text", "text": "回音壁已触发"}]}
    
    elif name == "zircon_evolve":
        return {"content": [{"type": "text", "text": "进化循环已触发"}]}
    
    return {"content": [{"type": "text", "text": f"未知工具: {name}"}]}

# === MCP 主循环 ===
def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        
        try:
            req = json.loads(line)
        except:
            print(mcp_error(None, -32700, "Parse error"))
            continue
        
        req_id = req.get("id")
        method = req.get("method", "")
        
        if method == "initialize":
            resp = mcp_response(req_id, {
                "protocolVersion": "2024-11-05",
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "zircon-evolution", "version": "1.0.0"}
            })
        elif method == "tools/list":
            resp = mcp_response(req_id, {"tools": TOOLS})
        elif method == "tools/call":
            params = req.get("params", {})
            tool_name = params.get("name", "")
            args = params.get("arguments", {})
            result = handle_tool(tool_name, args)
            resp = mcp_response(req_id, result)
        else:
            resp = mcp_error(req_id, -32601, "Method not found")
        
        print(resp)
        sys.stdout.flush()

if __name__ == "__main__":
    main()

File: core-dna/mcp-server.rs
/**
 * Zircon MCP Server — 标准 MCP 协议
 * 把 Rust 引擎变成标准 MCP 服务器
 * 任何支持 MCP 的 LLM 都能直接调用
 */

use std::io::{self, BufRead, Write};
use std::collections::HashMap;

// === MCP 标准响应 ===
fn mcp_response(id: Option<&serde_json::Value>, result: serde_json::Value) -> String {
    let mut resp = serde_json::json!({
        "jsonrpc": "2.0",
        "result": result,
    });
    if let Some(id) = id {
        resp["id"] = id.clone();
    }
    resp.to_string()
}

fn mcp_error(id: Option<&serde_json::Value>, code: i32, message: &str) -> String {
    let mut resp = serde_json::json!({
        "jsonrpc": "2.0",
        "error": {"code": code, "message": message},
    });
    if let Some(id) = id {
        resp["id"] = id.clone();
    }
    resp.to_string()
}

// === MCP 工具定义 ===
fn get_tools() -> serde_json::Value {
    serde_json::json!([
        {
            "name": "mutate",
            "description": "执行基因变异",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "domain": {"type": "string", "description": "变异维度"},
                    "change": {"type": "number", "description": "变异幅度"}
                },
                "required": ["domain", "change"]
            }
        },
        {
            "name": "evaluate",
            "description": "评估基因效果",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "gene_id": {"type": "string", "description": "基因ID"}
                },
                "required": ["gene_id"]
            }
        },
        {
            "name": "balance",
            "description": "计算洛书平衡度",
            "inputSchema": {"type": "object", "properties": {}}
        },
        {
            "name": "status",
            "description": "获取系统状态",
            "inputSchema": {"type": "object", "properties": {}}
        }
    ])
}

// === 主循环 ===
fn main() {
    let stdin = io::stdin();
    let mut stdout = io::stdout();
    
    for line in stdin.lock().lines() {
        let line = match line {
            Ok(l) => l,
            Err(_) => break,
        };
        
        let request: serde_json::Value = match serde_json::from_str(&line) {
            Ok(v) => v,
            Err(_) => {
                writeln!(stdout, "{}", mcp_error(None, -32700, "Parse error")).unwrap();
                continue;
            }
        };
        
        let id = request.get("id");
        let method = request.get("method").and_then(|m| m.as_str()).unwrap_or("");
        
        let response = match method {
            "initialize" => {
                mcp_response(id, serde_json::json!({
                    "protocolVersion": "2024-11-05",
                    "capabilities": {"tools": {}},
                    "serverInfo": {"name": "zircon-evolution", "version": "1.0.0"}
                }))
            }
            "tools/list" => {
                mcp_response(id, serde_json::json!({
                    "tools": get_tools()
                }))
            }
            "tools/call" => {
                let params = request.get("params").unwrap_or(&serde_json::json!({}));
                let tool_name = params.get("name").and_then(|n| n.as_str()).unwrap_or("");
                let args = params.get("arguments").unwrap_or(&serde_json::json!({}));
                
                let result = match tool_name {
                    "mutate" => {
                        let domain = args.get("domain").and_then(|d| d.as_str()).unwrap_or("unknown");
                        let change = args.get("change").and_then(|c| c.as_f64()).unwrap_or(0.0);
                        serde_json::json!({
                            "content": [{"type": "text", "text": format!("变异: {} ({})", domain, change)}]
                        })
                    }
                    "evaluate" => {
                        let gene_id = args.get("gene_id").and_then(|g| g.as_str()).unwrap_or("unknown");
                        serde_json::json!({
                            "content": [{"type": "text", "text": format!("评估: {}", gene_id)}]
                        })
                    }
                    "balance" => {
                        serde_json::json!({
                            "content": [{"type": "text", "text": "洛书平衡度计算"}]
                        })
                    }
                    "status" => {
                        serde_json::json!({
                            "content": [{"type": "text", "text": "系统状态正常"}]
                        })
                    }
                    _ => {
                        writeln!(stdout, "{}", mcp_error(id, -32601, "Method not found")).unwrap();
                        continue;
                    }
                };
                mcp_response(id, result)
            }
            _ => {
                mcp_error(id, -32601, "Method not found")
            }
        };
        
        writeln!(stdout, "{}", response).unwrap();
        stdout.flush().unwrap();
    }
}
